### ISSUE TITLE
Separate custom transaction life-cycle from automatic transaction

### ISSUE THREAD

**Issue by jahtalab** [elastic/apm-agent-rum-js#334]
Currently, transactions that are created by the user (custom transactions) and transactions that are created by the agent, live in the same space and can interfere with each other. This creates a number of issues:

* There can only be one active transaction at any give time
* Some transactions might be ended unexpectedly since another transaction is started.

To resolve these issues, it helps to think of the transactions in two different modes. One is automatic mode which means the agent controls the life-cycle of the transaction as well as creation of it's spans and the other one is manual mode in which case the agent does not create spans nor it detects the end of the transaction. we can add a property to the Transaction class that controls the mode and by default makes the transaction into manual mode.

With this approach we can address the following use-cases:
* Creating multiple transactions at the same time
* Capturing http requests as Transactions

The drawback of this approach is that custom transactions will no longer have automatic span creation (e.g. for XHR and fetch), to solve this we can give users the option to make a transaction in automatic mode.

WDYT, @vigneshshanmugam , @axw

**Comment by axw** [comment]
Sounds good to me.

Is there anything stopping us also using/implementing something along the lines of https://github.com/angular/zone.js/? With something like that we could have an active transaction per zone, with `apm.startSpan` operating on the current zone's active transaction.

**Comment by vigneshshanmugam** [comment]
Thanks Hamid for taking a stab at this. Seems super important and providing two modes at the same time seems like a good approach. 

> The drawback of this approach is that custom transactions will no longer have automatic span creation (e.g. for XHR and fetch), to solve this we can give users the option to make a transaction in automatic mode.

I feel like instead of doing this, We can provide events/some other mechanism to the users that would provide details of current span/marks that happen during the custom transaction. But these events only happen when we are not instrumenting the application using some other transaction(which is in automatic mode).

**Comment by jahtalab** [comment]
@axw , early versions of the RUM agent used zone.js, but we decided to remove it, the main reason was the bundle size (zone.js is not very light). However, we can probably implement a simpler variation of zones in the RUM agent, although, zone management is not trivial to implement, so I would like to defer it until we have enough use-cases to justify it.

@vigneshshanmugam , I thought about having individual events for our instrumentation, but I think that is more complicated for the users. IMO, having modes makes it easy for the user to choose at a high level, it's also an easier solution to implement.

**Comment by vigneshshanmugam** [comment]
Yeah sure, events would be a bit more complicated. I am worried about this scenario, When users turn the transaction in to automatic mode and there are currently 2 transactions running, How do we assign the corresponding automatically created spans to the which transaction.

We might have to force ourselves to log to the users that only one automatic transaction would be possible at a time in the application.

**Comment by axw** [comment]
@jahtalab fair enough, thanks for the background. A couple of alternative options come to mind: we could provide users with a means of registering their already-existing `zone.js`; or we could provide a separate bundle which layers on top. Anyways, we can of course defer this.

**Comment by jahtalab** [comment]
@v, I think the mode decision can only be made at the start of transaction, if the mode is automatic then we keep the current logic, i.e. end the old active automatic transaction and set the new transaction as the active. The logic for span creation is also the same as before for automatic mode, i.e. we only capture automatic spans (XHR, fetch) into the active transaction which by definition must be in automatic mode.

The only difference is in manual mode, first we don't end the transaction automatically and second we don't capture spans automatically as well. We might need to introduce a different method to create transaction in manual mode, instead of `startTransaction` which operates in automatic mode.
